File: app/src/main/java/com/lfork/a98620/lfree/data/source/IMDataRepository.java
package com.lfork.a98620.lfree.data.source;

import com.lfork.a98620.lfree.data.entity.User;
import com.lfork.a98620.lfree.data.source.local.IMLocalDataSource;
import com.lfork.a98620.lfree.data.source.remote.IMRemoteDataSource;
import com.lfork.a98620.lfree.data.source.remote.imservice.MessageService;
import com.lfork.a98620.lfree.data.source.remote.imservice.request.LoginListener;

import java.util.LinkedList;
import java.util.List;

/**
 */
public class IMDataRepository implements IMDataSource {

    private static IMDataRepository INSTANCE;

    private IMDataSource mRemoteDataSource, mLocalDataSource;

    private User mCachedUser;

//    private HashMap<Integer, Friend> mCachedFriendsMap;

    private List<User> mCachedUserList;

    private boolean mCachedIMIsDirty, mCachedFriendListIsDirty;

    public static final int USER_THIS = -1;

    private MessageService.MessageBinder binder;


    private IMDataRepository(IMDataSource mRemoteTRDataSource, IMDataSource mLocalDataSource) {
        this.mRemoteDataSource = mRemoteTRDataSource;
        this.mLocalDataSource = mLocalDataSource;
    }


    public static IMDataRepository getInstance() {
        if (INSTANCE != null) {
            return INSTANCE;
        } else {
            INSTANCE = new IMDataRepository(IMRemoteDataSource.getInstance(), IMLocalDataSource.getInstance());
            return INSTANCE;
        }
    }

    public static void destroyInstance() {
        IMRemoteDataSource.releaseInstance();
//        IMLocalDataSource.releaseInstance();
        INSTANCE = null;
    }

    public void setServiceBinder(MessageService.MessageBinder binder) {
        this.binder = binder;

    }

    public MessageService.MessageBinder getBinder() {
        return binder;
    }

    @Override
    public void login(User user, LoginListener listener) {
        mRemoteDataSource.login(user, new LoginListener() {
            @Override
            public void succeed(User user) {
                //如果是Android的话 这里记得要返回UI线程  使用Handler或者是Handler的封装
                mCachedUser = user;
                listener.succeed(user);
            }

            @Override
            public void failed(String log) {
                //如果是Android的话 这里记得要返回UI线程  使用Handler或者是Handler的封装
                listener.failed(log);
            }
        });

    }

    @Override
    public void logout(int userId, GeneralCallback<User> result) {
        new Thread(() -> {
            mRemoteDataSource.logout(userId, result);
        }).start();
    }

    @Override
    public synchronized void getChatUserList(GeneralCallback<List<User>> callback) {
        if (mCachedUserList != null) {
            callback.succeed(mCachedUserList);
            return;
        }
        mLocalDataSource.getChatUserList(new GeneralCallback<List<User>>() {
            @Override
            public void succeed(List<User> data) {
                mCachedUserList = data;
                callback.succeed(data);
            }

            @Override
            public void failed(String log) {
                callback.failed(log);
            }
        });
    }

    @Override
    public synchronized void addChatUser(User user, GeneralCallback<String> callback) {

    }

    @Override
    public synchronized void addChatUser(User user, boolean isExisted, GeneralCallback<String> callback) {
        if (isExisted) {
            if (mCachedUserList != null) {
                for (User u : mCachedUserList) {
                    if (user.getUserId() == u.getUserId()) {
                        mCachedUserList.remove(u);
                        addChatUserInLocal(u, callback);
                        return;
                    }
                }
            }
        }
        addChatUserInLocal(user, callback);
    }

    private synchronized void addChatUserInLocal(User user,  GeneralCallback<String> callback){
        mLocalDataSource.addChatUser(user, new GeneralCallback<String>() {
            @Override
            public void succeed(String data) {

                if (mCachedUserList == null) {
                    mCachedUserList = new LinkedList<>();
                }
                mCachedUserList.add(0, user);
                callback.succeed(data);
            }

            @Override
            public void failed(String log) {
                callback.failed(log);

            }
        });

    }

    @Override
    public synchronized void removeChatUser(int userId, GeneralCallback<List<User>> callback) {
        mLocalDataSource.removeChatUser(userId, new GeneralCallback<List<User>>() {
            @Override
            public void succeed(List<User> data) {
                for (int i = 0; i < mCachedUserList.size(); i++) {
                    User u = mCachedUserList.get(i);
                    if (u.getUserId() == userId) {
                        mCachedUserList.remove(i);
                        break;
                    }
                }
                callback.succeed(mCachedUserList);
            }

            @Override
            public void failed(String log) {
                callback.failed(log);
            }
        });

    }

    public synchronized boolean isUserExisted(int userId) {
        if (mCachedUserList != null) {
            for (User u : mCachedUserList) {
                if (userId == u.getUserId()) {
                    return true;
                }
            }
        }
        return false;
    }

//    public void getUserIndex(GeneralCallback<Integer> callback) {
//
//    }

//    @Override
//    public void getChatUserInfo(int userId, GeneralCallback<String> callback) {
//
//    }


}

File: app/src/main/java/com/lfork/a98620/lfree/data/imdata/IMDataRepository.kt
package com.lfork.a98620.lfree.data.imdata

import com.lfork.a98620.lfree.base.FreeApplication
import com.lfork.a98620.lfree.data.DataSource.GeneralCallback
import com.lfork.a98620.lfree.data.base.entity.User
import com.lfork.a98620.lfree.data.imdata.local.IMLocalDataSource
import com.lfork.a98620.lfree.data.imdata.remote.IMRemoteDataSource
import com.lfork.a98620.lfree.imservice.MessageService
import com.lfork.a98620.lfree.imservice.request.LoginListener
import java.util.*

/**
 * Created by 98620 on 2018/9/3.
 */

class IMDataRepository : IMDataSource {
    private val mRemoteDataSource = IMRemoteDataSource.getInstance()
    private val mLocalDataSource = IMLocalDataSource.getInstance()

    /**
     * 伪单例、不需要线程同步、懒汉模式
     */
    companion object {
        private var INSTANCE: IMDataRepository? = null
        fun getInstance(): IMDataRepository {
            if (INSTANCE == null) {
                INSTANCE = IMDataRepository()
            }
            return INSTANCE as IMDataRepository
        }

        fun destroyInstance() {
            IMRemoteDataSource.destroyInstance()
            IMLocalDataSource.destroyInstance()
            INSTANCE = null
        }
    }
    private var mCachedUser: User? = null

    /**
     * This variable has package local visibility so it can be accessed from tests.  重点来了，写成map的话不会报错
     */
    internal var mCachedUsers: MutableMap<String, User>? = null


    private var mCachedUsersIsDirty: Boolean = false

    var binder: MessageService.MessageBinder? = null
        private set

    fun setServiceBinder(binder: MessageService.MessageBinder) {
        this.binder = binder
    }

    override fun login(user: User, listener: LoginListener) {
        mRemoteDataSource.login(user, object : LoginListener {
            override fun succeed(user: User) {
                //如果是Android的话 这里记得要返回UI线程  使用Handler或者是Handler的封装
                mCachedUser = user
                listener.succeed(user)
            }

            override fun failed(log: String) {
                //如果是Android的话 这里记得要返回UI线程  使用Handler或者是Handler的封装
                listener.failed(log)
            }
        })

    }

    override fun logout(userId: Int, result: GeneralCallback<String>) {
        FreeApplication.executeThreadInDefaultThreadPool { mRemoteDataSource.logout(userId, result) }
    }

    @Synchronized
    override fun getChatUserList(callback: GeneralCallback<List<User>>) {
        if (mCachedUsers != null) {
            val users = ArrayList(mCachedUsers!!.values)
            users.sortWith(Comparator {           //小的排前面
                o1, o2 ->
                if (o1.timestamp > o2.timestamp) {
                    - 1
                } else {
                    1
                }
            })

            callback.succeed(users)
            return
        }
        mLocalDataSource.getChatUserList(object : GeneralCallback<List<User>> {
            override fun succeed(data: List<User>) {
                refreshCache(data)
                callback.succeed(data)
            }

            override fun failed(log: String) {
                callback.failed(log)
            }
        })
    }

    @Synchronized
    override fun addChatUser(user: User, callback: GeneralCallback<String>) {

    }

    @Synchronized
    override fun addChatUser(user: User, isExisted: Boolean, callback: GeneralCallback<String>) {
        if (isExisted) {
            if (mCachedUsers != null) {
                mCachedUsers!!.remove(user.userId.toString() + "")
                addChatUserInLocal(user, callback)
            }
        } else {
            addChatUserInLocal(user, callback)
        }
    }

    @Synchronized
    private fun addChatUserInLocal(user: User, callback: GeneralCallback<String>) {
        mLocalDataSource.addChatUser(user, object : GeneralCallback<String> {
            override fun succeed(data: String) {

                if (mCachedUsers == null) {
                    mCachedUsers = LinkedHashMap()
                }
                mCachedUsers!![user.userId.toString() + ""] = user
                callback.succeed(data)
            }

            override fun failed(log: String) {
                callback.failed(log)

            }
        })

    }

    @Synchronized
    override fun removeChatUser(userId: Int, callback: GeneralCallback<List<User>>) {
        mLocalDataSource.removeChatUser(userId, object : GeneralCallback<List<User>> {
            override fun succeed(data: List<User>) {

                if (mCachedUsers != null) {
                    mCachedUsers!!.remove(userId.toString() + "")
                    callback.succeed(ArrayList(mCachedUsers!!.values))
                } else {
                    callback.failed("没有数据")
                }


            }

            override fun failed(log: String) {
                callback.failed(log)
            }
        })

    }

    private fun refreshCache(users: List<User>) {
        if (mCachedUsers == null) {
            mCachedUsers = LinkedHashMap()
        }
        mCachedUsers!!.clear()
        for (user in users) {
            mCachedUsers!![user.userId.toString() + ""] = user
        }
        mCachedUsersIsDirty = false
    }


    @Synchronized
    fun isUserExisted(userId: Int): Boolean {
        if (mCachedUsers != null) {
            val user = mCachedUsers!![userId.toString() + ""]
            return user != null
        }
        return false
    }


}
